### DISCUSSION TITLE
Tree Data Grid x Parent/Child Relation Dataset x Failures on subsequent component mount

### DISCUSSION THREAD

**Discussion by tnaum-ms** [ghiscoding/slickgrid-universal#1655] [General]
I'm currently investigating an issue that arises when I mount my SlickGrid Tree Data as a React component for the second time. The problem occurs when I toggle between different data view formats in my application. Specifically, switching away from the Tree Data view and then returning to it results in a failure of the Tree Data Component.

The data set I’m working with follows a Parent/Child structure, where the `id` and `parentId` fields are of type `string`.
_(It would be helpful to include a note in the documentation indicating the preferred or faster data input method.)_

The SlickGrid Tree Data Grid throws the following error:

> Each data element must implement a unique 'id' property

Upon investigation, I discovered that on the second mount (not just a second render), one of the internal data mutation functions introduces duplicates into my data set. This causes the `ensureIdUniqueness` function to fail.

The issue seems to be related to the function located here:
https://github.com/ghiscoding/slickgrid-universal/blob/ce8b4000cf72fa78293e50fba886ae883c08f39b/packages/common/src/services/utilities.ts#L158

This specific loop appears to be introducing duplicates:
https://github.com/ghiscoding/slickgrid-universal/blob/ce8b4000cf72fa78293e50fba886ae883c08f39b/packages/common/src/services/utilities.ts#L173-L187

I'm continuing to investigate the root cause, but since this issue only occurs on the second mount, I haven’t yet identified any obvious side effects. If this is something you’ve encountered before, I would greatly appreciate any insights or suggestions you could share.

Thank you for your assistance.
Best, Tomasz

**Comment by ghiscoding** [comment]
Hmm I was hoping to never touch this code again, it took me near 6 months to get Tree Data working with all features that I wanted. I agree that calling this function more than once shouldn't impact its final array or should at the very least also provide the same output. However, it is important to note that the function does mutate the input array as it was mentioned in the JSDocs, but yes the result should always be the same and not add extra rows on the second run. But even then, the question would be why are you mounting it twice? Can't you make sure that you only mount the component once and avoid performing the same execution? I'm saying this because dealing with the Tree Data multiple time can be costly, I think I already mentioned that in your other discussion but... Side note, SlickGrid doesn't support Tree Data by itself, it can only deal with flat data, so the way I got it working is to keep copies of both the flat and hierarchical arrays and then keep them in sync whenever a CRUD operation comes in. So having to recreate both arrays (flat/hierarchical) multiple times is something you should avoid redoing too often because it can become costly in terms of performance

are you able to provide copies of the data? I think the first troubleshooting would be to see what happens when you run this function twice on the same array, I'm assuming the 2nd run modifies the array. If you can't provide the data, then you could try to troubleshoot it on your side by running the same function `unflattenParentChildArrayToTree()` twice.

oh and yeah I saw you mentioned Hello Team earlier, there's no team here, it's just me, welcome to the world of open source 😅

**Reply by tnaum-ms** [reply]
I looked into it in more detail and discovered the side effect. Now that you've pointed out the JSDoc, I realize it could have shortened my research... if I had read it in time! :) Nevertheless, it was enjoyable to debug and learn how everything works behind the scenes.

In the end, I realized that `unflattenParentChildArrayToTree` does mutate my input array, as it doesn't work on a copy of the data. Among other changes, it adds the `children` array:

https://github.com/ghiscoding/slickgrid-universal/blob/ce8b4000cf72fa78293e50fba886ae883c08f39b/packages/common/src/services/utilities.ts#L179-L181


This is the source of the issue. On a second run of `unflattenParentChildArrayToTree`, the same children are being pushed into the `children` array, leading to duplicates:

https://github.com/ghiscoding/slickgrid-universal/blob/ce8b4000cf72fa78293e50fba886ae883c08f39b/packages/common/src/services/utilities.ts#L182


I'm not sure what the best approach is to avoid issues on multiple runs of `unflattenParentChildArrayToTree` in the future:

- Using a `Set` would be problematic with objects.
- Filtering on insert by `id` only.
-Cleaning up on unmount.
- ...?


Regarding the double mount:
This relates to the application design. Users can choose between three views: table view, tree view, or text view. I aim to keep only one active at a time. I don't expect users to switch often; they'll likely choose their preferred view and stick with it.

By the way, I had a good laugh when I realized I initially started with "Hello Team." Guess that's a habit I'll have to unlearn in the wild world of open source :-D

**Reply by ghiscoding** [reply]
Maybe just a check to see if it's already in the children array? Something along these lines (untested). 
```ts
const idx = p[childrenPropName]?.findIndex(x => x.id === item.id);
if (idx < 0) {
  p[childrenPropName].push(item);
} else {
  p[childrenPropName][idx] = item; // replace existing one (probably equal to the same item)
}
```
I don't think I could use a Set because this must be a simple array, going with a Set would be at least a breaking change since it would be different on the user side (unless we convert it back to an array but that's extra work that I try to avoid for obvious perf reasons... mutating the original array is also for perf reasons)

BTW, there's another push here that would require the same kind of changes for the items in the root

https://github.com/ghiscoding/slickgrid-universal/blob/ce8b4000cf72fa78293e50fba886ae883c08f39b/packages/common/src/services/utilities.ts#L175-L177

**Reply by tnaum-ms** [reply]
After learning about the side effects on my data and the `children` array, I was able to work around it in my application code with something as simple as this:
![image](https://github.com/user-attachments/assets/abc6b627-d0a6-483b-b738-bab60c22bc6f)

Perhaps a workaround like this could be recommended for corner cases like mine, avoiding the need for another lookup in the codebase? While this might not help those who don't read the docs ;) it could be beneficial to add a mention in the tree data example source code.

By the way, I'm not too concerned about `roots` since it's a local variable that gets initialized as empty here:

https://github.com/ghiscoding/slickgrid-universal/blob/ce8b4000cf72fa78293e50fba886ae883c08f39b/packages/common/src/services/utilities.ts#L165

**Reply by ghiscoding** [reply]
> By the way, I'm not too concerned about `roots` since it's a local variable that gets initialized as empty here:

oh yes you're right on that one. I think my suggestion still makes sense though compare to your function, because your suggestion requires to again loop through every single rows of your dataset which is still not ideal on the perf side. 

I think an easier patch for you, until we implement my suggestion above, would be to simply keep a deep copy of your original array and simply reapply this copy every time you mount (it should be faster than deleting all children). For the deep copy, you can use some of my utils for this, the `extend` can work (which is `node-extend` and is the same implementation as jQuery `extend`). `const arrayCopy = extend(true, [], originalArray)`... but that keeps another copy in memory (not ideal). 
BUT wait, that's not actually a good suggestion if you do CRUD stuff on your data, then you're out of sync, so forget the suggestion :D

so up to you, at least you have more than 1 ways to go around the issue. I would still like to try my suggestion above and push a PR with this if it makes sense

**Reply by ghiscoding** [reply]
@tnaum-ms So I gave a try to my previous suggestion above and it works exactly like it should even after calling the `unflattenParentChildArrayToTree()` function multiple times. 

The way I've tested this was to first modify my unit tests to call `unflattenParentChildArrayToTree()` twice and as expected the array is getting duplicate items (the issue that you mentioned above). I then modified the code with my suggested fix and then called the unflatten function twice or even multiple times and it now always produce the same result which fixes the issue you've mentioned. 🎉 

The fix is in PR #1658 (merged) and will be released next weekend via a patch version (done).

Thanks a lot for all the troubleshooting that you've done, this helped quite a lot to identify the problem and bring a fix for this bug. 😃

**Reply by ghiscoding** [reply]
The fix is now available everywhere. Cheers

**Reply by tnaum-ms** [reply]
@ghiscoding Thank you very much for following up and for directly patching my code! I truly appreciate it. It made me aware of the fix you've rolled out. Fantastic!

**Reply by ghiscoding** [reply]
Welcome, it's always great to see some of my work being used on large project and company 😃 
I will surely use the extension in the future since I know there's a little bit of my work and effort in it 😉
